x/payment/types/wirepayfordata.go

```diff
@@ -122,9 +133,54 @@ func (msg *MsgWirePayForData) ValidateBasic() error {
 		return ErrNoMessageShareCommitments
 	}
 
+	if err := msg.ValidateAllSquareSizesCommitedTo(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateAllSquareSizesCommitedTo returns an error if the list of square sizes
+// committed to don't match all square sizes expected for this message size.
+func (msg *MsgWirePayForData) ValidateAllSquareSizesCommitedTo() error {
+	allSquareSizes := AllSquareSizes(int(msg.MessageSize))
+	committedSquareSizes := msg.committedSquareSizes()
+
+	if len(allSquareSizes) != len(committedSquareSizes) {
+		return ErrInvalidShareCommitments.Wrapf("length of all square sizes: %v must equal length of committed square sizes: %v", len(allSquareSizes), len(committedSquareSizes))
+	}
+
+	sort.Slice(allSquareSizes, func(i, j int) bool { return allSquareSizes[i] < allSquareSizes[j] })
+	sort.Slice(committedSquareSizes, func(i, j int) bool { return committedSquareSizes[i] < committedSquareSizes[j] })
+
+	if !isEqual(allSquareSizes, committedSquareSizes) {
+		return ErrInvalidShareCommitments.Wrapf("all square sizes: %v, committed square sizes: %v", allSquareSizes, committedSquareSizes)
+	}
 	return nil
 }
 
+// isEqual returns true if the given uint64 slices are equal
+func isEqual(a, b []uint64) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// commitedSquareSizes returns a list of square sizes that are present in a
+// message's share commitment.
+func (msg *MsgWirePayForData) committedSquareSizes() (squareSizes []uint64) {
+	for _, commit := range msg.MessageShareCommitment {
+		squareSizes = append(squareSizes, commit.K)
+	}
+	return squareSizes
```

Comment:
> why not pre-define the slice capacity here?

By default I favor readability over performance. When one needs to optimize for performance, they can evaluate with benchmarks and selectively make the existing readable code more performant where needed. I think the code above is _slightly_ more readable than the following but I understand this is subjective.

```golang
func (msg *MsgWirePayForData) committedSquareSizes() []uint64 {
	squareSizes := make([]uint64, 0, len(msg.MessageShareCommitment))
	...
}
```

> I think tendermint goes to the extent to include this in the linter.

You're right. Looks like:
- https://github.com/tendermint/tendermint/blob/f1835338d9f44d7bc215d84ff0c72a49c61897d3/.golangci.yml#L22
- https://golangci-lint.run/usage/linters/#prealloc

I'm in favor of enabling a linter like this to: (a) reduce PR feedback cycles and (b) build more codebase consistency. 

**Note**: the linter config says

```
    # IMPORTANT: we don't recommend using this linter before doing performance profiling.
    # For most programs usage of prealloc will be a premature optimization.
```

I attempted to enable locally but it didn't flag the instance in this PR 

```shell
$ golangci-lint run --verbose
...
INFO [lintersdb] Active 15 linters: [deadcode errcheck exportloopref gofumpt gosimple govet ineffassign misspell prealloc revive staticcheck structcheck typecheck unused varcheck]
...
// no issues reported
```

To avoid blocking this PR on this nit, will pre-allocate and create an issue to enable the linter (after confirming it works)

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/666/commits/30b8e8e66734a4f6a75e980f4a4ec895ec09328e

Reply:
Created https://github.com/celestiaorg/celestia-app/issues/676